extensions/cornerstone/src/components/ViewportWindowLevel/ViewportWindowLevel.tsx

```diff
@@ -0,0 +1,329 @@
+          );
+          const {
+            voi: displaySetVOI,
+            colormap: displaySetColormap,
+          } = viewportInfo.displaySetOptions[volumeIndex];
+          let colormap;
+
+          if (displaySetColormap) {
+            colormap =
+              csUtils.colormap.getColormap(displaySetColormap.name) ??
+              vtkColorMaps.getPresetByName(displaySetColormap.name);
+          }
+
+          const voi = !volumeIndex
+            ? viewportVoi ?? displaySetVOI
+            : displaySetVOI;
+
+          return {
+            modality,
+            volumeId,
+            volumeIndex,
+            voi,
+            histogram,
+            colormap,
+            step: scaling?.PET ? 0.05 : 1,
+          };
+        })
+        .filter(windowLevel => !!windowLevel.histogram);
+
+      return windowLevels;
+    },
+    [cornerstoneViewportService, getViewportVolumeHistogram]
+  );
+
+  const updateViewportHistograms = useCallback(() => {
+    setWindowLevels(prevWindowLevels =>
+      getWindowLevelsData(viewportIndex, prevWindowLevels)
+    );
+  }, [viewportIndex, getWindowLevelsData]);
+
+  const handleVOIModified = useCallback(
+    e => {
+      const { detail } = e;
+      const { volumeId, range } = detail;
+      const oldWindowLevel = windowLevels.find(wl => wl.volumeId === volumeId);
+
+      if (!oldWindowLevel) {
+        return;
+      }
+
+      const oldVOI = oldWindowLevel.voi;
+      const windowWidth = range.upper - range.lower;
+      const windowCenter = range.lower + windowWidth / 2;
+
+      if (
+        windowWidth === oldVOI.windowWidth &&
+        windowCenter === oldVOI.windowCenter
+      ) {
+        return;
+      }
+
+      const newWindowLevel = {
+        ...oldWindowLevel,
+        voi: {
+          windowWidth,
+          windowCenter,
+        },
+      };
+
+      setWindowLevels(
+        windowLevels.map(windowLevel =>
+          windowLevel === oldWindowLevel ? newWindowLevel : windowLevel
+        )
+      );
+    },
+    [windowLevels]
+  );
+
+  const debouncedHandleVOIModified = useCallback(
+    debounce(handleVOIModified, 100),
+    [handleVOIModified]
+  );
+
+  const handleVOIChange = useCallback(
+    (volumeId, voi) => {
+      const viewport = cornerstoneViewportService.getCornerstoneViewportByIndex(
+        viewportIndex
+      );
+
+      const newRange = {
+        lower: voi.windowCenter - voi.windowWidth / 2,
+        upper: voi.windowCenter + voi.windowWidth / 2,
+      };
+
+      viewport.setProperties({ voiRange: newRange }, volumeId);
+      viewport.render();
+    },
+    [cornerstoneViewportService, viewportIndex]
+  );
+
+  const handleOpacityChange = useCallback(
+    (_volumeIndex, _volumeId, opacity) => {
+      // const viewport = cornerstoneViewportService.getCornerstoneViewportByIndex(
+      //   viewportIndex
+      // );
+      // const properties = viewport.getProperties(volumeId);
+
+      console.warn(`TODO: update opacity: ${opacity}`);
```

Comment:
@sedghi As we discussed yesterday this branch needs to get merged with latest `master` and I need to take a look into the changes made to `opacityMapping` to see how this would work.